Rewrite sourcecatalog ingestion for flexibility

spenczar opened the pull request:
Deleted precovery.sourcecatalog.SourceExposure. Its functionality was purely internal to the sourcecatalog module, and it can be accomplished using a dictionary instead, which is less code weight.

Rewrote most of sourcecatalog. Instead of primarily operating on CSV files, it breaks that step into two phases. One phase is loading in a stream of SourceObservations. This could be from a CSV file, but could also be from a DataFrame, for example. The second phase is organizing the SourceObservations into chunks grouped by HEALPixel ("SourceFrames"), suitable for loading into the database. Those two phases are now contained in separate functions.

spenczar committed:
Rewrite sourcecatalog ingestion for flexibility

Deleted precovery.sourcecatalog.SourceExposure. Its functionality was
purely internal to the sourcecatalog module, and it can be
accomplished using a dictionary instead, which is less code weight.

Rewrote most of sourcecatalog. Instead of primarily operating on CSV
files, it breaks that step into two phases. One phase is loading in a
stream of SourceObservations. This could be from a CSV file, but could
also be from a DataFrame, for example. The second phase is organizing
the SourceObservations into chunks grouped by HEALPixel
("SourceFrames"), suitable for loading into the database. Those two
phases are now contained in separate functions.

spenczar committed:
Add a few tests of sourcecatalog

spenczar committed:
More tests of sourcecatalog bundling

akoumjian reviewed (approved):
I'll write up an implementation for datafram in etl repo and if we like it we can port it here afterwards

spenczar committed:
Remove unique constraint from fast_query index

The uniqueness constraint can't be obeyed if exposure data
is fed in across several passes, since rows are never updated
in our current design.

spenczar commented on precovery/frame_db.py:
```diff
@@ -505,9 +505,7 @@ def initialize_tables(self):
-            sq.Index(
-                "fast_query", "exposure_mjd_mid", "healpixel", "obscode", unique=True
-            ),
+            sq.Index("fast_query", "exposure_mjd_mid", "healpixel", "obscode"),
```
I'd like @moeyensj's eyes on this - do you foresee a big impact from removing `unique=True` from this index?

moeyensj replied:
I honestly don't have a good sense of the performance impact this change will have but it seems like a necessary one to enable the upcoming pipeline. The uniqueness constraint was originally added because it was just another test to check if the input HDF5 files were sorted correctly and that exposures were correctly grouped. We assumed up until that point that we would always have the files sorted.
